bindings/matrix-sdk-ffi/src/api.udl

```diff
@@ -112,57 +106,179 @@ interface Room {
     [Throws=ClientError]
     string? member_display_name(string user_id);
 
-    BackwardsStream? start_live_event_listener();
-    void stop_live_event_listener();
+    void add_timeline_listener(TimelineListener listener);
+
+    // Loads older messages into the timeline.
+    //
+    // Raises an exception if there are no timeline listeners.
+    [Throws=ClientError]
+    PaginationOutcome paginate_backwards();
+
+    // Removes the timeline.
+    //
+    // Timeline items cached in memory as well as timeline listeners are dropped.
+    void remove_timeline();
 
     [Throws=ClientError]
     void send(MessageEventContent msg, string? txn_id);
 };
 
-interface BackwardsStream {
-    sequence<AnyMessage> paginate_backwards(u64 count);
+callback interface TimelineListener {
+    void on_update(TimelineDiff update);
 };
 
-interface MessageEventContent {};
+// Should really be an enum
+interface TimelineDiff {
+    TimelineChange change();
 
-interface AnyMessage {
-    TextMessage? text_message();
-    ImageMessage? image_message();
-    NoticeMessage? notice_message();
-    EmoteMessage? emote_message();
+    [Self=ByArc]
+    sequence<TimelineItem>? replace();
+    [Self=ByArc]
+    InsertAtData? insert_at();
+    [Self=ByArc]
+    UpdateAtData? update_at();
+    u32? remove_at();
+    MoveData? move();
+    [Self=ByArc]
+    TimelineItem? push();
 };
 
-interface BaseMessage {
-    string id();
-    string body();
+enum TimelineChange {
+    "Replace",
+    "InsertAt",
+    "UpdateAt",
+    "RemoveAt",
+    "Move",
+    "Push",
+    "Pop",
+    "Clear",
+};
+
+dictionary InsertAtData {
+    u32 index;
+    TimelineItem item;
+};
+
+dictionary UpdateAtData {
+    u32 index;
+    TimelineItem item;
+};
+
+dictionary MoveData {
+    u32 old_index;
+    u32 new_index;
+};
+
+interface TimelineItem {
+    [Self=ByArc]
+    EventTimelineItem? as_event();
+    [Self=ByArc]
+    VirtualTimelineItem? as_virtual();
+};
+
+interface EventTimelineItem {
+    TimelineKey key();
     string sender();
+    TimelineItemContent content();
+    sequence<Reaction> reactions();
```

Comment:
Ideally, these would be aggregated reactions on the timeline API. (At a later date a separate method to request individual reactions for a specific event would be useful to show more details but we don't particularly need that in the timeline itself).

That said, looking at the checklist on the PR I think this is known :)

Reply:
Yeah, this doesn't work for pagination because there you don't have the full details. I think it should still be in the data model such that hovers over the symbols can be implemented easily. It will just not be available until you fetch all details of the event.